src/hotspot/share/gc/shenandoah/shenandoahMmuTracker.cpp

```diff
@@ -70,49 +93,111 @@ double ShenandoahMmuTracker::process_time_seconds() {
+    _most_recent_gcu = gc_time / (_active_processors * gc_cycle_duration);
+    double mutator_time = mutator_thread_time - _most_recent_mutator_time;
+    _most_recent_mutator_time = mutator_thread_time;
+    _most_recent_mu = mutator_time / (_active_processors * gc_cycle_duration);
+    log_info(gc, ergo)("At end of %s: GCU: %.1f%%, MU: %.1f%% for duration %.3fs",
+                       msg, _most_recent_gcu * 100, _most_recent_mu * 100, gc_cycle_duration);
+  }
 }
 
-ShenandoahMmuTracker::~ShenandoahMmuTracker() {
-  _mmu_periodic_task->disenroll();
-  delete _mmu_periodic_task;
+void ShenandoahMmuTracker::record_young(ShenandoahGeneration* generation, uint gcid) {
+  help_record_concurrent(generation, gcid, "Concurrent Young GC");
 }
 
-void ShenandoahMmuTracker::record(ShenandoahGeneration* generation) {
-  shenandoah_assert_control_or_vm_thread();
-  double collector_time_s = gc_thread_time_seconds();
-  double elapsed_gc_time_s = collector_time_s - _generational_reference_time_s;
-  generation->add_collection_time(elapsed_gc_time_s);
-  _generational_reference_time_s = collector_time_s;
+void ShenandoahMmuTracker::record_bootstrap(ShenandoahGeneration* generation, uint gcid, bool candidates_for_mixed) {
+  // Not likely that this will represent an "ideal" GCU, but doesn't hurt to try
+  help_record_concurrent(generation, gcid, "Bootstrap Old GC");
+  if (candidates_for_mixed) {
+    _doing_mixed_evacuations = true;
+  }
+  // Else, there are no candidates for mixed evacuations, so we are not going to do mixed evacuations.
+}
+
+void ShenandoahMmuTracker::record_old_marking_increment(ShenandoahGeneration* generation, uint gcid, bool old_marking_done,
+                                                        bool has_old_candidates) {
+  // No special processing for old marking
+  double duration = os::elapsedTime() - _most_recent_timestamp;
+  double gc_time, mutator_time;
+  fetch_cpu_times(gc_time, mutator_time);
+  double gcu = (gc_time - _most_recent_gc_time) / duration;
+  double mu = (mutator_time - _most_recent_mutator_time) / duration;
+  if (has_old_candidates) {
+    _doing_mixed_evacuations = true;
+  }
+  log_info(gc, ergo)("At end of %s: GC Utilization: %.1f%% for duration %.3fs (which is subsumed in next concurrent gc report)",
+                     old_marking_done? "last OLD marking increment": "OLD marking increment",
+                     _most_recent_gcu * 100, duration);
+}
+
+void ShenandoahMmuTracker::record_mixed(ShenandoahGeneration* generation, uint gcid, bool is_mixed_done) {
+  help_record_concurrent(generation, gcid, "Mixed Concurrent GC");
+}
+
+void ShenandoahMmuTracker::record_degenerated(ShenandoahGeneration* generation,
+                                              uint gcid, bool is_old_bootstrap, bool is_mixed_done) {
+  if ((gcid == _most_recent_gcid) && _most_recent_is_full) {
+    // Do nothing.  This is a redundant recording for the full gc that just completed.
```

Comment:
We have a redundant call here because some degenerated cycles become full cycles? Could we arrange for this to only be called once from each code path?

Reply:
It's complicated.  I'll put a todo statement here to suggest future refactoring.